File: backend/schemas/project.py
"""墨韵 - 项目相关 Schemas"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class ProjectCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=100, description="项目名称")
    genre: str = Field(default="", description="题材")
    theme: str = Field(default="", description="主题")
    tone: str = Field(default="", description="基调")
    background: str = Field(default="", description="故事背景")
    writing_style: str = Field(default="", description="写作风格")
    target_word_count: int = Field(default=100000, ge=10000, description="目标字数")
    author: str = Field(default="", description="作者名")
    # 场景级配置（sec = 单场景）
    scene_target_chars: int = Field(default=800, description="单场景目标字数")
    scenes_per_chapter: int = Field(default=5, description="每章节场景数")
    chapters_per_volume: int = Field(default=12, description="每卷章节数")


class ProjectUpdateRequest(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=100, description="项目名称")
    genre: Optional[str] = Field(default=None, description="题材")
    theme: Optional[str] = Field(default=None, description="主题")
    tone: Optional[str] = Field(default=None, description="基调")
    background: Optional[str] = Field(default=None, description="故事背景")
    writing_style: Optional[str] = Field(default=None, description="写作风格")
    target_word_count: Optional[int] = Field(default=None, ge=10000, description="目标字数")
    author: Optional[str] = Field(default=None, description="作者名")
    # 场景级配置
    scene_target_chars: Optional[int] = Field(default=None, description="单场景目标字数")
    scenes_per_chapter: Optional[int] = Field(default=None, description="每章节场景数")
    chapters_per_volume: Optional[int] = Field(default=None, description="每卷章节数")


class ProjectInfo(BaseModel):
    project_id: str
    name: str
    author: str = ""
    genre: str = ""
    theme: str = ""
    tone: str = ""
    background: str = ""
    writing_style: str = ""
    target_word_count: int = 0
    completion_rate: float = 0.0  # 0.0 ~ 1.0
    total_words: int = 0
    created_at: datetime
    updated_at: datetime
    # 场景级配置（sec = 单场景）
    scene_target_chars: int = 800
    scenes_per_chapter: int = 5
    chapters_per_volume: int = 12
    unit_label: str = "scene"


class ProjectListResponse(BaseModel):
    projects: list[ProjectInfo]
    total: int


class ProjectStatsResponse(BaseModel):
    """项目统计信息"""
    total_sections: int = 0
    completed_sections: int = 0
    total_words: int = 0
    chapter_count: int = 0
    volume_count: int = 0
    completion_rate: float = 0.0


# ─── Wizard 流程 Schemas ──────────────────────────────────────────

class BookIdeaRequest(BaseModel):
    """生成书名和创意的请求"""
    genre: str = Field(..., description="题材")
    tone: str = Field(default="", description="基调")
    background: str = Field(default="", description="故事背景")
    theme: str = Field(default="", description="主题")
    writing_style: str = Field(default="", description="写作风格")
    author: str = Field(default="", description="作者名")
    target_word_count: int = Field(default=50000, description="目标字数")


class BookIdeaResponse(BaseModel):
    """书名和创意响应"""
    name: str = Field(..., description="生成的书名")
    description: str = Field(..., description="创意描述")


class GenerateOutlineRequest(BaseModel):
    """生成大纲的请求"""
    genre: str = Field(..., description="题材")
    tone: str = Field(default="", description="基调")
    background: str = Field(default="", description="故事背景")
    theme: str = Field(default="", description="主题")
    writing_style: str = Field(default="", description="写作风格")
    author: str = Field(default="", description="作者名")
    target_word_count: int = Field(default=50000, description="目标字数")
    book_name: str = Field(default="", description="书名")
    book_description: str = Field(default="", description="创意描述")


class OutlineChapterInfo(BaseModel):
    """章节信息（场景级：每个章节包含多个场景）"""
    id: str
    name: str
    sections: int = 5  # 默认每章节5个场景
    scenes: int = 5   # 场景数量，与 sections 含义相同


class OutlineResponse(BaseModel):
    """大纲响应"""
    outline: str = Field(..., description="大纲内容（Markdown格式）")
    chapters: list[OutlineChapterInfo] = Field(default_factory=list, description="章节列表")


class ConfirmOutlineRequest(BaseModel):
    """确认大纲请求"""
    outline: str = Field(..., description="大纲内容（Markdown格式）")

File: frontend/src/stores/project.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/services/api'
import { API_ROUTES } from '@/shared/api/routes'
import type { ProjectInfo, ProjectCreateRequest } from '@/shared/api/types'

/** Store-level project type — ProjectInfo with defaults applied (required fields) */
export interface Project extends ProjectInfo {
  project_id: string
  name: string
  author: string
  genre: string
  tone: string
  background: string
  theme: string
  writing_style: string
  target_word_count: number
  completion_rate: number
  total_words: number
  created_at: string
  updated_at: string
}

export type CreateProjectParams = ProjectCreateRequest

export const useProjectStore = defineStore('project', () => {
  const projects = ref<Project[]>([])
  const currentProject = ref<Project | null>(null)
  const isLoading = ref(false)
  const isCreating = ref(false)
  const pendingGeneration = ref<{ filePath: string; prompt: string; promptType?: string; extraVars?: Record<string, string> } | null>(null)

  async function loadProjects() {
    isLoading.value = true
    try {
      const result = await api.get<{ projects: Record<string, unknown>[]; total: number }>(API_ROUTES.projects)
      projects.value = (result?.projects || []).map(normalizeProject)
    } finally {
      isLoading.value = false
    }
  }

  async function createProject(params: CreateProjectParams) {
    isCreating.value = true
    try {
      const raw = await api.post<Record<string, unknown>>(API_ROUTES.projects, {
        name: params.name || '新项目',
        author: params.author || '',
        genre: params.genre || '',
        tone: params.tone || '',
        background: params.background || '',
        theme: params.theme || '',
        writing_style: params.writing_style || '',
        target_word_count: params.target_word_count || 50000,
      })
      const project = normalizeProject(raw)
      projects.value.push(project)
      currentProject.value = project
      return project
    } finally {
      isCreating.value = false
    }
  }

  async function openProject(id: string) {
    const raw = await api.get<Record<string, unknown>>(API_ROUTES.projectDetail(id))
    const project = normalizeProject(raw)
    currentProject.value = project
    return project
  }

  async function deleteProject(id: string) {
    await api.delete(API_ROUTES.projectDetail(id))
    projects.value = projects.value.filter((p) => p.id !== id)
    if (currentProject.value?.id === id) {
      currentProject.value = null
    }
  }

  async function updateProject(id: string, data: Partial<Project>) {
    const raw = await api.put<Record<string, unknown>>(API_ROUTES.projectDetail(id), data)
    const updated = normalizeProject(raw)
    const index = projects.value.findIndex((p) => p.id === id)
    if (index !== -1) {
      projects.value[index] = updated
    }
    if (currentProject.value?.id === id) {
      currentProject.value = updated
    }
    return updated
  }

  function closeProject() {
    currentProject.value = null
  }

  function setPendingGeneration(val: { filePath: string; prompt: string; promptType?: string; extraVars?: Record<string, string> } | null) {
    pendingGeneration.value = val
  }

  function calculateCompletion(project: Project): number {
    if (!project.target_word_count) return 0
    return Math.round((project.total_words / project.target_word_count) * 100)
  }

  return {
    projects,
    currentProject,
    isLoading,
    isCreating,
    loadProjects,
    createProject,
    openProject,
    deleteProject,
    updateProject,
    closeProject,
    calculateCompletion,
    pendingGeneration,
    setPendingGeneration,
  }
}, {
  persist: {
    storage: localStorage,
    pick: ['currentProject'],
  },
})

function normalizeProject(p: Record<string, unknown>): Project {
  const pid = (p.project_id as string) || (p.id as string)
  return {
    id: pid,
    project_id: pid,
    name: (p.name as string) || '',
    author: (p.author as string) || '',
    genre: (p.genre as string) || '',
    tone: (p.tone as string) || '',
    background: (p.background as string) || '',
    theme: (p.theme as string) || '',
    writing_style: (p.writing_style as string) || '',
    target_word_count: (p.target_word_count as number) || 0,
    completion_rate: (p.completion_rate as number) || 0,
    total_words: (p.total_words as number) || 0,
    created_at: (p.created_at as string) || '',
    updated_at: (p.updated_at as string) || '',
    scene_target_chars: (p.scene_target_chars as number) || 800,
    scenes_per_chapter: (p.scenes_per_chapter as number) || 5,
    chapters_per_volume: (p.chapters_per_volume as number) || 12,
    unit_label: (p.unit_label as string) || 'scene',
  }
}
